cardano-cli/src/Cardano/CLI/Shelley/Run/Query.hs

```diff
@@ -375,13 +435,87 @@ writeLedgerState mOutFile qState@(SerialisedLedgerState serLedgerState) =
     Just (OutputFile fpath) ->
       handleIOExceptT (ShelleyQueryCmdWriteFileError . FileIOError fpath)
         $ LBS.writeFile fpath $ unSerialised serLedgerState
- where
-   decodeLedgerState
-     :: SerialisedLedgerState era
-     -> Either LBS.ByteString (LedgerState era)
-   decodeLedgerState (SerialisedLedgerState (Serialised ls)) =
-     first (const ls) (decodeFull ls)
 
+writeStakeSnapshot :: forall era ledgerera. ()
+  => ShelleyLedgerEra era ~ ledgerera
+  => Era.Era ledgerera
+  => FromCBOR (LedgerState era)
+  => Hash StakePoolKey
+  -> SerialisedLedgerState era
+  -> ExceptT ShelleyQueryCmdError IO ()
+writeStakeSnapshot poolId qState =
+  case decodeLedgerState qState of
+    -- In the event of decode failure print the CBOR instead
+    Left bs -> firstExceptT ShelleyQueryCmdHelpersError $ pPrintCBOR bs
+
+    Right ledgerState -> do
+      -- Convert the hash string into a KeyHash for use by the ledger
+      hk <- KeyHash <$> hashFromStringAsHex (filter (/= '"') (show poolId))
```

Comment:
We _should_ be able to use `serialiseToRawBytesHexText` here instead of `hashFromStringAsHex (filter (/= '"') (show poolId)`

Reply:
This seems to not work as `serialiseToRawBytesHexText` returns `Text` but I need a `Maybe (Hash (ADDRHASH (Crypto ledgerera)) (VerKeyDSIGN (DSIGN (Crypto ledgerera))))`

Reply:
[1mdiff --git a/cardano-cli/src/Cardano/CLI/Shelley/Run/Query.hs b/cardano-cli/src/Cardano/CLI/Shelley/Run/Query.hs[m
[1mindex 10c695228..a61afcf84 100644[m
[1m--- a/cardano-cli/src/Cardano/CLI/Shelley/Run/Query.hs[m
[1m+++ b/cardano-cli/src/Cardano/CLI/Shelley/Run/Query.hs[m
[36m@@ -438,22 +438,17 @@[m [mwriteLedgerState mOutFile qState@(SerialisedLedgerState serLedgerState) =[m
 [m
 writeStakeSnapshot :: forall era ledgerera. ()[m
   => ShelleyLedgerEra era ~ ledgerera[m
[31m-  => Era.Era ledgerera[m
[32m+[m[32m  => Era.Crypto ledgerera ~ StandardCrypto[m
   => FromCBOR (LedgerState era)[m
   => PoolId[m
   -> SerialisedLedgerState era[m
   -> ExceptT ShelleyQueryCmdError IO ()[m
[31m-writeStakeSnapshot poolId qState =[m
[32m+[m[32mwriteStakeSnapshot (StakePoolKeyHash hk) qState =[m
   case decodeLedgerState qState of[m
     -- In the event of decode failure print the CBOR instead[m
     Left bs -> firstExceptT ShelleyQueryCmdHelpersError $ pPrintCBOR bs[m
 [m
     Right ledgerState -> do[m
[31m-      let x = hashFromStringAsHex (filter (/= '"') (show poolId))[m
[31m-      -- Convert the hash string into a KeyHash for use by the ledger[m
[31m-      hk <- KeyHash <$> x[m
[31m-        & hoistMaybe (ShelleyQueryCmdPoolIdError poolId)[m
[31m-[m
       -- Ledger State[m
       let (LedgerState snapshot) = ledgerState[m
 [m
[36m@@ -764,6 +759,7 @@[m [mobtainLedgerEraClassConstraints[m
   -> ((Ledger.ShelleyBased ledgerera[m
       , ToJSON (LedgerState era)[m
       , FromCBOR (LedgerState era)[m
[32m+[m[32m      , Era.Crypto ledgerera ~ StandardCrypto[m
       ) => a) -> a[m
 obtainLedgerEraClassConstraints ShelleyBasedEraShelley f = f[m
 obtainLedgerEraClassConstraints ShelleyBasedEraAllegra f = f[m

Reply:
```diff
[1mdiff --git a/cardano-cli/src/Cardano/CLI/Shelley/Run/Query.hs b/cardano-cli/src/Cardano/CLI/Shelley/Run/Query.hs[m
[1mindex 10c695228..a61afcf84 100644[m
[1m--- a/cardano-cli/src/Cardano/CLI/Shelley/Run/Query.hs[m
[1m+++ b/cardano-cli/src/Cardano/CLI/Shelley/Run/Query.hs[m
[36m@@ -438,22 +438,17 @@[m [mwriteLedgerState mOutFile qState@(SerialisedLedgerState serLedgerState) =[m
 [m
 writeStakeSnapshot :: forall era ledgerera. ()[m
   => ShelleyLedgerEra era ~ ledgerera[m
[31m-  => Era.Era ledgerera[m
[32m+[m[32m  => Era.Crypto ledgerera ~ StandardCrypto[m
   => FromCBOR (LedgerState era)[m
   => PoolId[m
   -> SerialisedLedgerState era[m
   -> ExceptT ShelleyQueryCmdError IO ()[m
[31m-writeStakeSnapshot poolId qState =[m
[32m+[m[32mwriteStakeSnapshot (StakePoolKeyHash hk) qState =[m
   case decodeLedgerState qState of[m
     -- In the event of decode failure print the CBOR instead[m
     Left bs -> firstExceptT ShelleyQueryCmdHelpersError $ pPrintCBOR bs[m
 [m
     Right ledgerState -> do[m
[31m-      let x = hashFromStringAsHex (filter (/= '"') (show poolId))[m
[31m-      -- Convert the hash string into a KeyHash for use by the ledger[m
[31m-      hk <- KeyHash <$> x[m
[31m-        & hoistMaybe (ShelleyQueryCmdPoolIdError poolId)[m
[31m-[m
       -- Ledger State[m
       let (LedgerState snapshot) = ledgerState[m
 [m
[36m@@ -764,6 +759,7 @@[m [mobtainLedgerEraClassConstraints[m
   -> ((Ledger.ShelleyBased ledgerera[m
       , ToJSON (LedgerState era)[m
       , FromCBOR (LedgerState era)[m
[32m+[m[32m      , Era.Crypto ledgerera ~ StandardCrypto[m
       ) => a) -> a[m
 obtainLedgerEraClassConstraints ShelleyBasedEraShelley f = f[m
 obtainLedgerEraClassConstraints ShelleyBasedEraAllegra f = f[m

```

Reply:
```diff
diff --git a/cardano-cli/src/Cardano/CLI/Shelley/Run/Query.hs b/cardano-cli/src/Cardano/CLI/Shelley/Run/Query.hs
index 10c695228..a61afcf84 100644
--- a/cardano-cli/src/Cardano/CLI/Shelley/Run/Query.hs
+++ b/cardano-cli/src/Cardano/CLI/Shelley/Run/Query.hs
@@ -438,22 +438,17 @@ writeLedgerState mOutFile qState@(SerialisedLedgerState serLedgerState) =
 
 writeStakeSnapshot :: forall era ledgerera. ()
   => ShelleyLedgerEra era ~ ledgerera
-  => Era.Era ledgerera
+  => Era.Crypto ledgerera ~ StandardCrypto
   => FromCBOR (LedgerState era)
   => PoolId
   -> SerialisedLedgerState era
   -> ExceptT ShelleyQueryCmdError IO ()
-writeStakeSnapshot poolId qState =
+writeStakeSnapshot (StakePoolKeyHash hk) qState =
   case decodeLedgerState qState of
     -- In the event of decode failure print the CBOR instead
     Left bs -> firstExceptT ShelleyQueryCmdHelpersError $ pPrintCBOR bs
 
     Right ledgerState -> do
-      let x = hashFromStringAsHex (filter (/= '"') (show poolId))
-      -- Convert the hash string into a KeyHash for use by the ledger
-      hk <- KeyHash <$> x
-        & hoistMaybe (ShelleyQueryCmdPoolIdError poolId)
-
       -- Ledger State
       let (LedgerState snapshot) = ledgerState
 
@@ -764,6 +759,7 @@ obtainLedgerEraClassConstraints
   -> ((Ledger.ShelleyBased ledgerera
       , ToJSON (LedgerState era)
       , FromCBOR (LedgerState era)
+      , Era.Crypto ledgerera ~ StandardCrypto
       ) => a) -> a
 obtainLedgerEraClassConstraints ShelleyBasedEraShelley f = f
 obtainLedgerEraClassConstraints ShelleyBasedEraAllegra f = f
```

Reply:
Yep.  Worked.  Thanks!